core/src/main/java/jenkins/model/labels/LabelAutoCompleteSeeder.java

```diff
@@ -0,0 +1,59 @@
+package jenkins.model.labels;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Utility class for taking the current input value and computing a list of potential terms to match against the
+ * list of defined labels.
+ *
+ * @since TODO
+ */
```

Comment:
```suggestion
 */
@Restricted(NoExternalUse.class)
```

since I do not see it being used in any of the downstream PRs and I would expect it to be called only from `LabelExpression.autoComplete`.

Reply:
(plus imports of course)

Reply:
I'm confused. First I get asked to move it out of the class as a public type so it can be reused, and now you want me to flag it to prevent use.

Reply:
https://github.com/jenkinsci/jenkins/pull/4774#discussion_r440044574 for reference. Indeed there are a few plugins which have copied this class. The question is whether they actually needed to, or if they could call `LabelExpression.autoComplete` instead. If there is a legitimate need to refer to `LabelAutocompleteSeeder` directly, it could be unrestricted if and when there is a downstream PR demonstrating its proper use.

Reply:
Sorry for the mixup, just trying to avoid introducing new public APIs without a clear need.

Reply:
I don't really see why anyone would need to use it. I'd personally be in favour to move it back as a private nested implementation detail class.

Reply:
Assuming that a quick review of the existing copies suggests that they could use the simple `autoComplete` method, then yes a private nested implementation class would be fine. I see a usage from test sources though it was not clear if that was necessary or not. (Note that `@Restricted` is ignored in `src/test/java/**/*.java`. It is designed only to prevent improper use of types in _production_ code.)

Reply:
I'll do an org search and check, but I very much suspect all cases copied the seeder because they also copied the autocomplete method, which used the seeder.
As for the test, I think the nested class was package-private to make those tests work. I'm not super convinced that unit tests for the seeder make sense though, given that it's just an implementation detail. But it was there, so I kept it, and will continue to keep it until asked otherwise.

Reply:
Each of the cases found by the org search from https://github.com/jenkinsci/jenkins/pull/4774#discussion_r440044574 is a plugin that copied the doAutoCompleteXXX() method, and also copied the seeder - some as a nested class, some as a separate one. Think I'll just apply the `@Restricted` as suggested - it's less work than moving it back and adapting the unit test again too.

Reply:
Done